src/nvim/os/shell.c

```diff
@@ -309,19 +309,44 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
   dbuf->len += nread;
 }
 
+char *_output = NULL;
+
 /// Continue to append data to last screen line.
 ///
 /// @param output       Data to append to screen lines.
 /// @param remaining    Size of data.
 /// @param new_line     If true, next data output will be on a new line.
-static void append_to_screen_end(char *output, size_t remaining, bool new_line)
+static void append_to_screen_end(void *source, char *output, size_t remaining, bool new_line)
 {
+  static void *last_source = NULL;
   // Column of last row to start appending data to.
   static colnr_T last_col = 0;
+  char skip_msg[] = { '.', '\0', '\0' };
 
   size_t off = 0;
   int last_row = (int)Rows - 1;
 
+  if (last_source != source) {
+    last_source = source;
+  } else {
+    // count the lines
+    size_t lines = 0;
+    for (char *p = output; p < (output + remaining); p++) {
+      if (*p == NL) {
+        lines++;
+      }
+    }
```

Comment:
An idea for a more sophisticated "throttle trigger", might be the keep a count of the pending `thread_events` per-queue and then maybe ignore them until processing catches up. 

But maybe the arbitrary limit  (this PR) is good enough: it's probably not useful for anyone to see more than 1000 lines blasted in the pager, in _any_ UI (TUI or GUI). So the approach I was thinking here was to "snip" the middle of long outputs, and print the last N lines (as well as the first N lines).